OpenSearch: Migrate help article number-of-shards-and-indexes-in-elasticsearch

anelook opened the pull request:
# What changed, and why it matters

Migrated https://help.aiven.io/en/articles/2413410-number-of-shards-and-indexes-in-elasticsearch

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+Optimal number of shards
+========================
+
+OpenSearch has a default shard count limit per index (1024). Aiven does not place any additional restrictions on the number of shards that you can use for your OpenSearch service. However, maintaining higher count of shards comes with the performance cost, and can make the cluster less efficient. In this article we'll give advice on how to determine number of shards you need.
```
Do we have any doc that explains what shards are?

```suggestion
OpenSearch® has a default shard count limit per index (1024) and Aiven does not place any additional restrictions on the number of shards that you can use for your OpenSearch service. However, maintaining higher count of shards comes with the performance cost, and can make the cluster less efficient. In this article we'll give advice on how to determine number of shards you need.
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+
+OpenSearch has a default shard count limit per index (1024). Aiven does not place any additional restrictions on the number of shards that you can use for your OpenSearch service. However, maintaining higher count of shards comes with the performance cost, and can make the cluster less efficient. In this article we'll give advice on how to determine number of shards you need.
+
+Number of shards depends heavily on the amount of data you have.
```
```suggestion
The number of suggested shards depends heavily on the amount of data you have.
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+These suggestions are only indicative - optimal values depend heavily on
+your usage pattern and anticipated growth of data in OpenSearch.
+
+You can change number of shards without losing your data, but this
```
```suggestion
You can change the number of shards without losing your data, but this
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+your usage pattern and anticipated growth of data in OpenSearch.
+
+You can change number of shards without losing your data, but this
+process will require a brief downtime when index is rewritten.
```
```suggestion
process requires a brief downtime while the index is rewritten.
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+You can change number of shards without losing your data, but this
+process will require a brief downtime when index is rewritten.
+
+Performance numbers
```
Why not Performance implications? I feel Performance numbers is not very informative?

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+Performance numbers
+-------------------
+
+Having a large number of shards affect performance you get
```
```suggestion
Having a large number of shards affect your OpenSearch performance
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+-------------------
+
+Having a large number of shards affect performance you get
+out from OpenSearch. Some rough numbers from three-node Aiven
```
```suggestion
out from OpenSearch. Some benchmark results from three-node Aiven
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+
+Having a large number of shards affect performance you get
+out from OpenSearch. Some rough numbers from three-node Aiven
+OpenSearch business-8 cluster:
```
```suggestion
OpenSearch ``business-8`` cluster:
```

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+out from OpenSearch. Some rough numbers from three-node Aiven
+OpenSearch business-8 cluster:
+
+-  1 000 shards: no visible effect in OpenSearch performance.
```
What's the testing baseline? How much data? a benchmark without reference data is somehow less useful

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+
+-  1 000 shards: no visible effect in OpenSearch performance.
+
+-  10 000 shards is already quite a lot - creating new shards starts to
```
"is already quite a lot" <- i feel this is not quantitative, can we be more precise (or remove it)?

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+-  10 000 shards is already quite a lot - creating new shards starts to
+   take longer and longer time. Variance in performance grows.
+
+-  15 000 shards: creating new shards takes significantly longer time,
```
Can we give more direct results? instead of "significantly"?

ftisiot commented on docs/products/opensearch/concepts/shards-number.rst:
```diff
@@ -0,0 +1,53 @@
+   times (20x longer than mean). Similarly, variance in search
+   performance grows significantly.
+
+Aiven for OpenSearch takes a snapshot once every hour. With 10 000 shards
```
Should this be a warning?

ftisiot commented on docs/products/opensearch/concepts/index-usage.rst:
```diff
@@ -0,0 +1,36 @@
+indexes for very long time (years?), consider per-week or per-month
+indexes instead.
+
+It is not recommended to create a new index per customer/project/entity if
```
```suggestion
It is **not** recommended to create a new index per customer/project/entity if
```

ftisiot commented on docs/products/opensearch/concepts/index-usage.rst:
```diff
@@ -0,0 +1,36 @@
+Index usage
```
Can we make the title more clear about the content? Like 

"Indexes design" or "indexes usefulness" or something along the lines of understanding if an index is useful?

anelook replied:
I renamed it to "When to create a new index"

ftisiot reviewed (changes requested):
Good content!
I made a couple of comments, and have a major doubt: the benchmarking we are defining in the optimal number of shards doesn't provide reference and is pretty vague in terms of results. Should we conduct a more formal testing?

Also, final note, should the "Optimal # of shards" article being in the HowTo with a title "Calculate the Optimal # of shards
"?

lornajane commented:
I'm a little concerned about the long-term work-in-progress here. Could we rebase and try to merge what we know is good to go, and then iterate to add more?

anelook commented:
@lornajane, this was delayed because I needed to re-do the benchmarks and was looking into that. However, happy to cleanup and merge what is ready by now.

anelook commented:
I removed the benchmark section and added the fixes where there was a review feedback.
Created a ticket to add benchmarking separately - https://github.com/aiven/devportal/issues/898

ftisiot reviewed (approved):
LGTM
